### ISSUE TITLE
Exec/JavaExec Task can execute java processes in the background.	

### ISSUE THREAD

**Issue by bmuschko** [gradle/gradle#1367]
Original issue: https://issues.gradle.org/browse/GRADLE-1254

**Highly voted issue: 59**

### Expected Behavior

Introduce a configuration that allows the `Exec` and `JavaExec` task to execute the operation in the background.

### Current Behavior

The task waits until the executed process is finished.

### Context

Being able to decouple process execution from the execution of other tasks in the build.

**Comment by oehme** [comment]
This should be trivial once we have the asynchronous work API that @ghale is currently working on. The task will still wait for the action to complete, but other parallel-safe actions will be allowed to run at the same time.

**Comment by bassmake** [comment]
Is there any workaround for current gradle version?

**Comment by eskatos** [comment]
@oehme 
> This should be trivial once we have the asynchronous work API that @ghale is currently working on. The task will still wait for the action to complete, but other parallel-safe actions will be allowed to run at the same time.

I think there's some confusion here. The upcoming asynchronous work API won't allow to start a background process *around* other tasks. This issue is about just that, starting a background process, then running other tasks like tests, then shutting down that background process.

@bassmake 
> Is there any workaround for current gradle version?

Not that I know of. There's a community plugin (https://github.com/johnrengelman/gradle-processes ) that aim at addressing this need but it looks unmaintained.

**Comment by bassmake** [comment]
@eskatos 
yeah, I tried that. I also tried solution mentioned here: http://stackoverflow.com/a/38528039/1860580
but looks like `task.execute()` is no longer available. Is there some dirty way to execute task from another task?

**Comment by eskatos** [comment]
@bassmake nope. Please 👍 this issue if you want it to be prioritized and/or consider contributing.

**Comment by oehme** [comment]
Why do you want to execute a task from another task? There is `Project.exec()` which you could call from a Thread if that solves your use case.

**Comment by bassmake** [comment]
@oehme 
Thanks that helped a lot. I posted my solution here: http://stackoverflow.com/a/42557798/1860580

**Comment by psxpaul** [comment]
I created a plugin to do this a few months ago, for anyone looking for this functionality. I'd appreciate some feedback.
https://github.com/psxpaul/gradle-execfork-plugin

**Comment by eriwen** [comment]
@ghale Is the [Worker API](https://docs.gradle.org/current/userguide/custom_tasks.html#worker_api) *the* answer here? If so, do we have sufficient docs that show how one can do this?

**Comment by big-guy** [comment]
@eriwen I think there are a few things here, just "run in the background" isn't clear enough.

1) You have an `Exec`/`JavaExec` task that is long running and you'd like to do other work in the same project at the same time.  The Worker API would be the way to do this.  We should/could make this change internally. At the end of the task, the process should have exited (unless there's some kind of reuse).
2) You have a custom task that uses `project.exec {}` or `project.javaexec {}` to do work and you'd like to do other work in the same project at the same time (and/or have automatic resource constraints if you're forking a lot or reuse JVMs).  You can replace `project.javaexec {}` today with the worker API.  You can't currently inject any services into the worker context, so it's hard to do `project.exec {}` without dropping down to the Java ProcessBuilder API. At the end of the task, the process should have exited (unless there's some kind of reuse).
3) You have a custom/`Exec`/`JavaExec` task that starts up a service when you run integration tests. You'd like this to run in the background so you can also run integration tests.  This doesn't require the Worker API, but needs something like the deployment API that we've got some ideas about.
4) You have a custom/`Exec`/`JavaExec` task that starts up a service and then you manually clickety-click on it like `PlayRun` works.  Maybe you also want things to automatically restart when running in continuous build. This needs the deployment API too.

**Comment by ghale** [comment]
So, I think there are a bunch of different use cases in this space:

1. I want to run a task that executes some java command, but does not wait for the command to finish.  Gradle build finishes, the command stays running in the background.
1. Same as above, other tasks run while the command is running, the Gradle build finishes and the command also exits.
1. Same as above, except I run Gradle in continuous mode and the command stays running across builds, then exits when the build session ends.
1. Same as all above, except it's a non-java command.
1. I want to run a task that starts a java "daemon", I send requests to that daemon from that task (and potentially from other tasks in the build).  Gradle build finishes, the daemon stays running in the background.

There's probably other variations of the use cases above, too.  My understanding of the original issue here is that it was referring to 1 and/or 2.  The Worker API is really only designed to address 5 and is designed for a case where the implementation of the command can be refactored to a "unit of work" implementation.  In other words, you need some degree of control over the implementation on the server side and it's geared for handling requests that are part of the build execution.  Having said that, there are probably ways you could accomplish the other use cases with the worker API, but you would likely be jamming a square peg into a round hole.  For instance, could you find a way to start a web server in the background using the Worker API?  Yeah, probably.  Does that really make sense?  Probably not.

**Comment by yeikel** [comment]
I have a use case where I stumbled to this issue that I think that may help to document this issue. 

Consider the following  : 

1 - Build the backend
2 - Launch it
3 - Build the UI 
4 - Launch it
4 - Run the end to end tests  


```
task end2end(type: NpmTask) {
    description 'Builds and runs the API and the UI'
    dependsOn('::backend:api::build')
    dependsOn('::backend:api::bootRun')   -> It gets stuck here
    dependsOn('::frontend:ui::build')
    dependsOn('::frontend:ui::serve')
    dependsOn(npmInstall)
    args = ['test']
}
```

**Comment by DonkeyAfterDark** [comment]
I want to launch a task in the background, then go about doing other things without opening up a new Terminal window.

**Comment by binkley** [comment]
My team's use case is:
1) Bring up a Spring Boot REST service in the background
2) Run Postman (Newman) API tests in the foreground against the service
3) Shut down the service when tests complete

**Comment by olegshtch** [comment]
I got a very dirty solution to do it:

```groovy
import java.util.concurrent.atomic.AtomicReference

final AtomicReference serviceRef = new AtomicReference()

class SpawnJavaExec extends JavaExec {
	
	@Input
	AtomicReference spawnRef 
	
	@TaskAction
	@Override
	void exec() {
		setJvmArgs(getJvmArgs())
		def f = super.getClass().superclass.superclass.getDeclaredField('javaExecHandleBuilder')
		f.accessible = true
		def javaExecHandleBuilder = f.get(this)
		def execHandle = javaExecHandleBuilder.build()
		spawnRef.set execHandle.start()
	}
}

class AbortJavaExec extends DefaultTask {
	@Input
	AtomicReference spawnRef
	
	@TaskAction
	void abort() {
		spawnRef.get().abort()
		spawnRef.get().waitForFinish()
	}
}

task spawnService(type:SpawnJavaExec) {
	classpath = configurations.testRuntimeClasspath
	
	main = 'main.BaseApplication'
	
	spawnRef = serviceRef
}

task abortService(type:AbortJavaExec) {
	dependsOn spawnService
	
	spawnRef = serviceRef
}
```

**Comment by michaelcordero** [comment]
task runAppInBackground(type: Exec) {
    workingDir "$projectDir"
    commandLine "./gradlew", "run", "&"
    standardOutput = new ByteArrayOutputStream()
}

**Comment by immo-huneke-zuhlke** [comment]
Michael, I think you don't need the '&' to detach the background process.

Unfortunately the spawn plugin appears to be no longer maintained since version 0.8.2, created on 21 November 2016 - its originator (wiredforcode.com) no longer has a web site. So there's no documentation on the correct use of the plugin.

In case it helps, here is a snippet of my build.gradle file, which starts up a service provider as a background process in order to run PACT consumer-driven contract tests against it:
```
task startProvider(type: SpawnProcessTask, dependsOn: 'assemble') {
        command "java -Dspring.profiles.active=pactVerify -jar ${jar.archivePath}"
        ready "Started EasyNotesAppApplicationKt"
}

task stopProvider(type: KillProcessTask) {}

pact {
    serviceProviders {
        'easy-notes-app' {
                        port = 8989
                        startProviderTask = startProvider
                        terminateProviderTask = stopProvider
            // Only load the pacts from the broker if the start tasks from the command line include pactVerify
            if ('pactVerify' in gradle.startParameter.taskNames) {
                stateChangeUrl = url('http://localhost:8989/setup')
                stateChangeUsesBody = false // defaults to true
                hasPactsFromPactBrokerWithTag('http://localhost:80', 'test')
//                hasPactsFromPactBrokerWithTag('http://35.242.180.87:80', 'test')
            }
        }
    }
}
```

**Comment by joffrey-bion** [comment]
I have a use case similar to @binkley's (which falls under category 3 in @big-guy 's list).
I'm writing a multiplatform websocket client, and I'd like to start a websocket server so that I can run integration tests for all implementations.

Was there any progress on this issue? How should I go about this?

**Comment by jaro-sagan** [comment]
Hi @joffrey-bion, I do not think that there is some support for background processes in Gradle yet, but check this https://github.com/hesch/gradle-execfork-plugin it may help you with what you need to do.

**Comment by olegshtch** [comment]
Update of dirty solution for 7.3.3

```groovy
package ru.everestcollect.gradle

import java.util.concurrent.atomic.AtomicReference

public class SpawnJavaExec extends JavaExec {
	
	@Internal
	final AtomicReference spawnRef = new AtomicReference()
	
	@TaskAction
	@Override
	void exec() {
		setJvmArgs(getJvmArgs())
		def javaExecAction = getExecActionFactory().newJavaExecAction()
		def f = super.getClass().superclass.superclass.getDeclaredField('javaExecSpec')
		f.accessible = true
		def javaExecSpec = f.get(this)
		javaExecSpec.copyTo(javaExecAction)
		def m = super.getClass().superclass.superclass.getDeclaredMethod('getEffectiveExecutable')
		m.accessible = true		
		javaExecAction.executable = m.invoke(this)
		def execHandle = javaExecAction.build()
		spawnRef.set execHandle.start()
	}
}


```

**Comment by mainzelM** [comment]
The "dirty solution" works for me, I just had to add sourceSets.main.output to the classpath to be able to access the code of the current project.

**Comment by atatadev** [comment]
Probably late on the topic, but faced today and come up with wokers + build service solution.
### The setup
 I need deploy few external dependencies (using cli tool) before executing integrationTests and after it's ended I need to stop it.
 
### The solution
 - Build service where I run blocking deployment with CLI using execOperations in constructor:
 ```kotlin 
abstract class StackDeploymentService @Inject constructor(
  private val execOperations: ExecOperations,
) : BuildService<StackDeploymentService.Params>, AutoCloseable {

  interface Params : BuildServiceParameters {
    val devstackExecutableLocation: DirectoryProperty
    val devstackYml: RegularFileProperty
    val deploymentFile: RegularFileProperty
    val overrides: Property<String>
  }

  init {
    with(parameters) {
      execOperations.exec { commandLine("start", "here") }
    }

  }

  override fun close() {
    execOperations.exec { commandLine("stop", "here") }
  }
}
```
- Use task to deploy it by requesting BuildService above with WorkerExecutor:
```kotlin
abstract class DeployStackService : DefaultTask() {

    @get:ServiceReference("web")
    abstract val stackDeployment: Property<StackDeploymentService>
    @Inject
    abstract fun getWorkerExecutor(): WorkerExecutor

    @TaskAction
    fun deploy() {
        getWorkerExecutor().noIsolation()
            .submit(DeploymentWorker::class.java) {
                deployment.set(stackDeployment)
            }
    }


    abstract class DeploymentWorker : WorkAction<DeploymentWorker.Parameters> {
        interface Parameters: WorkParameters {
            abstract val deployment: Property<StackDeploymentService>
        }

        override fun execute() {
            //this will start cli
            parameters.deployment.get()
        }
    }
}
```
- in my  integrationTest task add dependency to deployment task:
```kotlin
tasks.integrationTest {
    dependsOn(tasks.deployStack)
    val service = gradle.sharedServices
        .registrations
        .getByName("devstack")
        .service
        .map { it as StackDeploymentService }
    service.map { it.getSomethingFromThisService }
}
```
In this case my deployment process, even though is blocking, is executed in background and doesn't block other tasks.
Also, it's automatically stopped, once all tasks that depend on the build service are finished
